src/OpenTelemetry.Contrib.Instrumentation.Remoting/Implementation/TelemetryDynamicSink.cs

```diff
@@ -0,0 +1,237 @@
+
+                // Do we already have an incoming activity?
+                // Existing activity might be started by an instrumentation layer higher up the
+                // stack. E.g. if we are using http as a transport for remoting, and we have
+                // instrumented incoming http request, then that instrumentation happens before
+                // our "ProcessMessageStart" and we just need to attach to that existing activity.
+                var outerActivity = Activity.Current;
+                if (outerActivity != null)
+                {
+                    ourActivity = RemotingActivitySource.StartActivity(ActivityInName, ActivityKind.Server);
+                }
+                else
+                {
+                    // We don't have an existing incoming activity. Start a brand new one ourselves.
+
+                    ourActivity = RemotingActivitySource.StartActivity(ActivityInName, ActivityKind.Server, activityParentContext.ActivityContext);
+                }
+
+                SetStartingActivityAttributes(ourActivity, reqMsg);
+            }
+        }
+
+        public void ProcessMessageFinish(IMessage replyMsg, bool bCliSide, bool bAsync)
+        {
+            // This will be called:
+            // 1) On the server after we are done processing the call and are returning back to client
+            //    => Current activity should be the "ActivityIn", we need to stop it
+            //
+            // 2) On the client, when the server call returns
+            //    => Current activity should be the "ActivityOut", we need to stop it as well
+
+            try
+            {
+                if (this.options.Filter?.Invoke(replyMsg) == false)
+                {
+                    return;
+                }
+            }
+            catch (Exception e)
+            {
+                RemotingInstrumentationEventSource.Log.MessageFilterException(e);
+                return;
+            }
+
+            var act = Activity.Current;
+            if (act == null)
+            {
+                RemotingInstrumentationEventSource.Log.NullActivity(nameof(TelemetryDynamicSink), nameof(this.ProcessMessageFinish));
+                return;
+            }
+
+            bool validClientActivity = bCliSide && act.OperationName == ActivityOutName;
+            bool validServerActivity = !bCliSide && act.OperationName == ActivityInName;
+            if (validClientActivity || validServerActivity)
+            {
+                if (replyMsg is IMethodReturnMessage returnMsg)
+                {
+                    if (returnMsg.Exception == null)
+                    {
+                        act.SetStatus(Status.Ok);
+                    }
+                    else
+                    {
+                        act.SetStatus(Status.Unknown);
+
+                        var e = returnMsg.Exception;
+                        act.SetTag(AttributeExceptionType, e.GetType().FullName);
```

Comment:
FYI you can use the [RecordException extension](https://github.com/open-telemetry/opentelemetry-dotnet/blob/f1ebf05ecb3e96d2069f105aa9a585ac81008552/src/OpenTelemetry.Api/Trace/ActivityExtensions.cs#L120) here.

Main benefit to that is it will add the details as an event which is what is called for [in the spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/exceptions.md).

Reply:
Thanks! I _did_ read that page before and completely missed the point about the whole point about the Event, thought it was about span attributes... Fixed now.